### PR TITLE
[FLINK-24325][connectors/elasticsearch] Create Elasticsearch 6 Sink

### PR DESCRIPTION
**Pull request by alpreu**
## What is the purpose of the change

This pull request creates a Elasticsearch 6 sink based on the new unified sink API (FLIP-143). 

**bf4244c and 30313bd need no review as they will be part of https://github.com/apache/flink/pull/17374.**

## Brief change log

  - To ease development shared code between Elasticsearch 6 and 7 connectors was extracted into the `flink-connector-elasticsearch-base` module
  - The Elasticsearch 5 compatible connector was removed
  - Many tests were migrated to the JUnit Jupiter API

## Verifying this change

This change is already covered by existing tests, such as:
* ElasticsearchDynamicSink - ITCase/FactoryTest/Test (for the respective ES versions)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs , also further documentation to be done as a separate task)

### CODE REVIEW

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.function.Function;
+
+/**
+ * {@link BulkRequestConsumerFactory} is used to bridge incompatible Elasticsearch Java API calls
+ * across different Elasticsearch versions.
+ */
+@Internal
+interface BulkRequestConsumerFactory
+        extends Function<
+                        RestHighLevelClient, BiConsumer<BulkRequest, ActionListener<BulkResponse>>>,
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java:39]: Do you gain anything by extending `Function`? I'd rather go with an explicit `create`.
An alternative cut may be to actually add an explicit `BulkRequestConsumer` (or handler?) class and use `SerializableFunction<RestHighLevelClient,  BulkRequestConsumer>` as the factory.

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java [hunk 2]
```diff
@@ -23,31 +24,21 @@
...
- *     .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)
- *     .build();
- * }</pre>
+ * Base builder to construct a {@link ElasticsearchSink}.
  *
  * @param <IN> type of the records converted to Elasticsearch actions
  */
 @PublicEvolving
-public class ElasticsearchSinkBuilder<IN> {
+public abstract class ElasticsearchSinkBuilderBase<IN> implements Serializable {
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java:41]: For base builders, you usually need to capture the type as well `ElasticsearchSinkBuilderBase<IN, B extends ElasticsearchSinkBuilderBase<IN>>` in case you want to add ES6 / 7 specific methods later.

Comment (AHeise) [reply flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java:41]: Why is the builder serializable?

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java [hunk 3]
```diff
@@ -80,15 +76,16 @@
      * Sets the emitter which is invoked on every record to convert it to Elasticsearch actions.
      *
      * @param emitter to process records into Elasticsearch actions.
-     * @return {@link ElasticsearchSinkBuilder}
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java:83]: replace everything with `@return this builder`. You usually link only to external types. In this case, the return type is actually a specific builder and not just base; so it's inaccurate.

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/RequestIndexer.java [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
  *
- *    http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/RequestIndexer.java:10]: Can you create a hotfix with all the license header changes (and double-check that yours is correct).

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBaseITCase.java [hunk 5]
```diff
@@ -165,8 +161,8 @@ private void runTest(
             @Nullable MapFunction<Long, Long> additionalMapper)
             throws Exception {
         final ElasticsearchSink<Tuple2<Integer, String>> sink =
-                ElasticsearchSink.builder()
-                        .setHosts(HttpHost.create(ES_CONTAINER.getHttpHostAddress()))
+                getSinkBuilder()
+                        .setHosts(HttpHost.create(getElasticsearchHttpHostAddress()))
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBaseITCase.java:165]: Can we move setting the host to `getSinkBuilder`? Then we have only one abstract method.

File: flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * Builder to construct an Elasticsearch 6 compatible {@link ElasticsearchSink}.
+ *
+ * <p>The following example shows the minimal setup to create a ElasticsearchSink that submits
+ * actions on checkpoint or the default number of actions was buffered (1000).
+ *
+ * <pre>{@code
+ * ElasticsearchSink<String> sink = new Elasticsearch6SinkBuilder<String>()
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java:30]: Let's change that to `ElasticsearchSink<String> sink = new Elasticsearch6Sink.builder()` to make that consistent with Kafka/Pulsar. Probably deserves a separate commit. Builder ctor should not be public.

File: flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ *     .setHosts(MY_ELASTICSEARCH_HOSTS)
+ *     .setEmitter(MY_ELASTICSEARCH_EMITTER)
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java:32]: In general, I think we should have more specific examples for the main entry points: There should be an example emitter and we should give an example host (the full URL). FYI, @fapaul

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchConnectorOptions.java [hunk 8]
```diff
@@ -153,19 +119,9 @@
...
-        DISABLED,
-        CONSTANT,
-        EXPONENTIAL
-    }
-
-    private ElasticsearchConnectorOptions() {}
+    public static final ConfigOption<DeliveryGuarantee> DELIVERY_GUARANTEE_OPTION =
+            ConfigOptions.key("sink.delivery-guarantee")
+                    .enumType(DeliveryGuarantee.class)
+                    .defaultValue(DeliveryGuarantee.AT_LEAST_ONCE)
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchConnectorOptions.java:125]: This is different from the default to DataStream. I'd probably favor at-least-once anyways, but we should align here.

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java [hunk 9]
```diff
@@ -81,21 +106,18 @@ public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
...
+                        createKeyExtractor());
 
-        final Elasticsearch7SinkBuilder<RowData> builder = new Elasticsearch7SinkBuilder<>();
+        final ElasticsearchSinkBuilderBase<RowData> builder = builderSupplier.get();
         builder.setEmitter(rowElasticsearchEmitter);
         builder.setHosts(config.getHosts().toArray(new HttpHost[0]));
         builder.setDeliveryGuarantee(config.getDeliveryGuarantee());
         builder.setBulkFlushMaxActions(config.getBulkFlushMaxActions());
-        builder.setBulkFlushMaxSizeMb((int) (config.getBulkFlushMaxByteSize() >> 20));
+        builder.setBulkFlushMaxSizeMb((int) (config.getBulkFlushMaxByteSize().getBytes() >> 20));
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java:120]: `getMebiBytes`

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java [hunk 10]
```diff
@@ -45,22 +48,44 @@
  * logical description.
  */
 @Internal
-final class Elasticsearch7DynamicSink implements DynamicTableSink {
+class ElasticsearchDynamicSink implements DynamicTableSink {
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java:51]: Does this mean we get rid of version specific ESDynamicSink in table? :D

**Review by AHeise — COMMENTED** [review]
Looks pretty good already. I especially like the code reduction (a large chunk is ES5 ofc). Could you please double-check if we didn't lose any tests in ES6/7?

**Comment by AHeise** [comment]
Btw it's hard to see if all comments of @fapaul have been addressed. Please always respond to feedback, so it's easier to follow what has been improved already. I use 👍 reaction in case of agreement instead of "yes, you are right, I'm incorporating as you said". Everything else should be a proper response.

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java (hunk 10 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSink.java:51]: Yes :)

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/IndexGeneratorTest.java [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+                                "log_ts",
+                                "log_date",
+                                "order_timestamp",
+                                "log_time",
+                                "local_datetime",
+                                "local_date",
+                                "local_time",
+                                "note",
+                                "status")
+                        .collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/IndexGeneratorTest.java:67]: 👍

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/IndexGeneratorTest.java [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+/** Suite tests for {@link IndexGenerator}. */
+public class IndexGeneratorTest {
+
+    private List<RowData> rows;
+    private List<String> fieldNames;
+    private List<DataType> dataTypes;
+
+    @BeforeEach
+    public void prepareData() {
```
[the remark this answers is not in the document]
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/IndexGeneratorTest.java:53]: 👍

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java (hunk 1 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java:39]: I'm not quite sure I understand, SerializableFunction also extends Function

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java (hunk 2 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBase.java:41]: Regarding the base builder I initially tried the type capture but could not get `setEmitter` to work with in this setup. If you could help me out there we can try to change it

The serializable was an accident, I'll remove it.

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBaseITCase.java (hunk 5 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBaseITCase.java:165]: No because we still need it for creating the RestHighlevelClient

File: flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java (hunk 6 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/connector/elasticsearch/sink/Elasticsearch6SinkBuilder.java:30]: There is no `Elasticsearch6Sink` anymore, just `ElasticsearchSink` in the base module so I'm not sure how we can keep this pattern in this case

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSinkBaseITCase.java [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+
+    abstract ElasticsearchDynamicSinkFactoryBase getDynamicSinkFactory();
+
+    abstract GetRequest createGetRequest(String index, String id);
+
+    abstract TestContext getPrefilledTestContext(String index);
+
+    abstract String getConnectorSql(String index);
+
+    abstract long getTotalSearchHits(SearchHits searchHits);
```
[the remark this answers is not in the document]
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchDynamicSinkBaseITCase.java]: After taking some time thinking and talking about this I did not come to find a better solution right now

File: flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java (hunk 1 above)
Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/connector/elasticsearch/sink/BulkRequestConsumerFactory.java]: In the end I opted for the interface with `create`

File: flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBaseTest.java [hunk 14]
```diff
@@ -19,27 +19,41 @@
...
+                                .setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE),
+                        createMinimalBuilder()
+                                .setBulkFlushBackoffStrategy(FlushBackoffType.CONSTANT, 1, 1),
+                        createMinimalBuilder()
+                                .setConnectionUsername("username")
+                                .setConnectionPassword("password"));
+
+        return DynamicTest.stream(
+                validBuilders,
+                ElasticsearchSinkBuilderBase::toString,
```
Comment (AHeise) [inline flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBaseTest.java:55]: Here we should actually have a proper name for the specific case. So we could use `Map<String, B> validBuilders` and then `put("At-least-once builder", createMinimalBuilder().setDeliveryGuarantee(DeliveryGuarantee.AT_LEAST_ONCE)`

Comment (alpreu) [reply flink-connectors/flink-connector-elasticsearch-base/src/test/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilderBaseTest.java:55]: As discussed offline, we keep the toString() as it shows all the builders fields as the test description

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure

**Comment by alpreu** [comment]
@flinkbot run azure
